src/aead/gcm.rs

```diff
@@ -116,3 +230,35 @@ struct GCM128_CONTEXT {
     H_unused: Block,
     key: GCM128_KEY,
 }
+
+enum Implementation {
+    CLMUL,
+
+    #[cfg(target_arch = "arm")]
+    NEON,
+
+    Fallback,
+}
+
+#[inline]
+fn detect_implementation(cpu: cpu::Features) -> Implementation {
+    if (cpu::intel::FXSR.available(cpu) && cpu::intel::PCLMULQDQ.available(cpu))
+        || cpu::arm::PMULL.available(cpu)
+    {
+        return Implementation::CLMUL;
+    }
```

Comment:
In the future SGX port, if we don't detect `Implementation::CLMUL` then we need to fail here.